### ISSUE TITLE
Открытие сохраненного проекта (сессии) - вкладки

### ISSUE THREAD

**Issue by SchDen** [Bashka/vim_prj#2]
Сессия не востанавливает вкладки. При открытии появляется только одна вкладка (имеется введу вкладки создаваемые через :tabnew). 
Добавил команду set sessionoptions=blank,buffers,curdir,folds,help,resize,tabpages,winsize
 не дало результатов

**Comment by Bashka** [comment]
Реализовал. Обновите плагин. Учтите, что сохраняются только те вкладки, в которых открыты файлы (не пустые или временные буферы).

**Comment by SchDen** [comment]
Попробовал обновление - зависания не происходит, но и открытия проекта тоже.
Вот экран, после выбора проекта и нажатия enter
![2015-06-16 0 32 01](https://cloud.githubusercontent.com/assets/1484977/8171241/28715b42-13bf-11e5-9355-93f4f7f88d52.png)

**Comment by SchDen** [comment]
Замечу, что в силу определенных предпочтений, ставил плагин vim_prj через простой клонирование в папку bundle, возможно это как то связано..
Еще пара моментов:
- при открытии vim через терминал, постоянно отображается ошибка:
  ![2015-06-16 0 39 48](https://cloud.githubusercontent.com/assets/1484977/8171388/3a368d60-13c0-11e5-9b1c-8bef5c7492cb.png)
  Причем появляется она вне зависимости от присутствия в каталоге проекта.
- В терминальном vim (VIM - Vi IMproved 7.3 (2010 Aug 15, compiled Aug 24 2013 18:58:47)) - он перемещается в папку с проектом, а в MacVim не происходит даже этого.

**Comment by Bashka** [comment]
_Попробовал обновление - зависания не происходит, но и открытия проекта тоже.
Вот экран, после выбора проекта и нажатия enter_
Обратите внимание, что если вы открываете проект, который ранее еще не открывали, то есть вы не открывали в этом проекте ни одного файла (нет текущего файла проекта), то при перехода к этому проекту из vim_start будет вновь запущено стартовое меню. Происходит это в связи с тем, что при открытии проекта открывается пустой буфер, а для vim_start это сигнал отрисовать меню. Вы можете проверить это, открыв проект с помощью vim_start и набрав команду :echo getcwd(), при этом на экране отобразится адрес корневого каталога проекта, что означает переход в проект. Далее нажмите клавишу "e", чтобы закрыть меню выбора проекта и откройте один из файлов проекта. Последующий выход из проекта с сохранением сессии и повторный вход в проект будет сопровождаться открытием этого файла, а не отрисовкой меню проекта.

**Comment by SchDen** [comment]
Я понимаю момент про стартовый экран, что нет файлов в проекте открытых, поэтому он отображается. Дело в том, что когда я нахожусь не в папке с проектом и перехожу к нему из стартовой странице мое положение в редакторе не изменяется. Проверяю через плагин nerdtree - может это не показатель, но если я делаю это из терминального vim, все срабатывает (условно), - я оказываюсь в папке с проектом.

**Comment by SchDen** [comment]
Про условность - не происходит открытия файлов проекта, т.е я закрывал с определенными файлами и вкладками - но не один из файлов, не открыт - постоянно открывается стартовая страница. Произошло вроде после обновления vim_start. 
После же обновления начала появляться ошибка, может дело в этом:
![2015-06-16 23 42 26](https://cloud.githubusercontent.com/assets/1484977/8194103/0c79ca14-1482-11e5-8a85-8c89736f738c.png)

**Comment by Bashka** [comment]
К сожалению, я не могу повторить у себя этот баг. Я постараюсь установить MacVim, для тестирования.

**Comment by prtngn** [comment]
Mac OS Yosemite. vim 7.3
Вылетает точно такая же ошибка. Правда я только начинаю изучать vim и вообще не понимаю что делаю =) Мой .vimrc

```
filetype off
set rtp=~/.vim/bundle/vim_lib
call vim_lib#sys#Autoload#init('~/.vim', 'bundle')

Plugin 'vim_lib'
Plugin 'vim_plugmanager'
Plugin 'vim_prj'
" Другие плагины

filetype indent plugin on
```

**Comment by Bashka** [comment]
Я протестирую это, как только доберусь до дома.

**Comment by F1asher086** [comment]
Всем доброго времени суток.
Давно наткнулся на вашу статью на Хабре, но опробовать вашу работу только. Уж сильно захотелось сделать из Vim'a IDE. Но при использовании vim_prj, я столкнулся с одной проблемкой. Я не могу понять как vim_prj сохраняет сессию, ведь при выходе из Vim приходится закрывать каждый буфер по отдельности. И в итоге в сессии сохраняется один единственный буфер, который был закрыт последним. 
Я далеко не опытный пользователь Vim и, может быть, не знаю, что можно выйти, закрыв все буферы сразу. Гуглил способ, но ничего не нашел. Прошу помощи.
(Не знал, куда лучше написать. Статье на хабре уже месяц, поэтому вряд ли кто-то туда заглянет. Поэтому написал сюда.)
Заранее благодарю за помощь.

**Comment by SchDen** [comment]
@F1asher086 попробуй :qa 
http://stackoverflow.com/a/11828573/2660839
У самого схожая проблема, не было времени разбираться

**Comment by F1asher086** [comment]
@SchDen Огромнейшее спасибо!

**Comment by SchDen** [comment]
@F1asher086 помогло?

**Comment by F1asher086** [comment]
@SchDen Да, теперь проект открывается в том состоянии, в котором был

**Comment by F1asher086** [comment]
@SchDen Только окошко NEDTree не остается, а так даже вкладки сохраняются

**Comment by SchDen** [comment]
@F1asher086 если требуется постоянно держать открым окошко NERD Tree, можно сделать как тут написано http://stackoverflow.com/questions/1759737/auto-open-nerdtree-in-vim, по поводу того, чтобы сохранить там состояние, которые было до закрытия не знаю.
Вы используете терминальный vim?

**Comment by F1asher086** [comment]
@SchDen Да, пользую терминальный vim. Вписал в .vimrc то, что советуют по ссылке. Теперь при входе в vim открывается NERDTree, но не работает vim_start. Пробовал строчку переставлять в разные места в файле -- не помогло: [ver.1](http://pastebin.com/tXKiQXGX) [ver.2](http://pastebin.com/1nqvmbDH) [ver.3](http://pastebin.com/2ypgL05d)
Также при открытии в директории проекта (созданного с vim_prj) проект не открывается, Открывается только NERDTree.
Не большая беда. Могу и через команду открывать. Хотя было бы неплохо забиндить ее на какую-нибудь клавишу.

**Comment by Bashka** [comment]
День добрый, F1asher086. Все верно, для сохранения проекта можно использовать команду :qa, которая закроет все буферы разом, что позволит их сохранить в сессии.
Касаемо окна NERDTree, которое не открывается при восстановлении проекта, можно решить эту проблему очень просто. Дело в том, что мои плагины используют событийную модель взаимодействия. Это означает, что можно вызвать некую функцию или команду при появлении некоторого события (на пример, после загрузки сессии проекта). Это можно использовать для открытия NERDTree. Просто добавьте в конец вашего .vimrc файла запись:
silent! echo vim_lib#sys#Publisher#
function! g:NERDTreeShowHidden(...) " {{{
  exe 'NERDTreeToggle'
  " Делаем активным рабочее окно, а не окно NERDTree
  exe winnr('#') . 'wincmd w'
endfunction " }}}
call vim_lib#sys#Publisher#.new().listen('VimPrjLoadSession', function('g:NERDTreeShowHidden'))

**Comment by Bashka** [comment]
День добрый, SchDen. Я помню про вашу проблему, связанную с использованием vim_prj на Mac OS. Сейчас очень занят написанием документации под плагины и Vim в целом (планируется большой проект). Постараюсь обязательно повторить вашу проблему у себя в ближайшее время.

**Comment by SchDen** [comment]
@Bashka да, хотелось наладить работу под Mac. Постепенно появляются новые ньюансы, и даже идеи для новых плагинов, буду писать по мере возможности.

**Comment by SchDen** [comment]
@Bashka и еще интересно было бы ваше мнение по поводу проекта https://github.com/neovim/neovim ,  по моим наблюдениям довольно активно ведется работа над ним и комьюнити растет хорошо

**Comment by Bashka** [comment]
@SchDen мнения как такового нет. Большой проект, хорошие идеи, активное развитие. Остается пожелать удачи участникам и ждать что из этого получится.

**Comment by F1asher086** [comment]
@Bashka Благодарю за помощь. Теперь все работает как надо.

**Comment by Bashka** [comment]
Появилась такая идея. Чтобы на события плагинов реагировать было проще неискушенным в VimScript пользователям, могу добавить такую логику: если в каталогах ~/.vim/listeners и Proj/.vim/listeners есть файлы с Vim скриптами, имя которых соответствует имени целевого события, то они будут выполняться при возникновении события.

То есть для обработки события VimPrjLoadSession достаточно будет создать файл ~/.vim/listeners/VimPrjLoadSession.vim и добавить в него:
NERDTreeToggle
winnr('#') . 'wincmd w'

Кажись так будет проще и удобнее.

**Comment by SchDen** [comment]
@Bashka думаю пригодиться
